### PR TITLE
construct a SSAGraph at the beginning.

### PR DESCRIPTION
**Pull request by panyx0718**

### CODE REVIEW

File: paddle/fluid/framework/details/multi_devices_graph_builder.cc [hunk 1]
```diff
@@ -186,19 +187,70 @@ size_t MultiDevSSAGraphBuilder::GetAppropriateDeviceID(
...
+    std::unique_ptr<ir::Graph> graph) const {
+  // Give the topology sort order and rebuild the graph structure.
+  std::vector<ir::Node *> sorted_ops = SortOpsAndDelayOptimizeOp(*graph);
+  auto nodes = graph->ReleaseNodes();
+  ir::Graph &result = *graph;
 
   for (auto &node : nodes) {
-    if (node->NodeType() == ir::Node::Type::kVariable) {
+    if (node->NodeType() == ir::Node::Type::kVariable && node->Var()) {
       all_vars_.emplace(node->Name(), node->Var());
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/multi_devices_graph_builder.cc:251]: Can you ensure that all the `node_name` are unique?

Comment (panyx0718) [reply paddle/fluid/framework/details/multi_devices_graph_builder.cc:251]: It does the same thing as before. So if the previous implementation ensure, then the change also ensures it.

**Review by chengduoZH — APPROVED** [review]

File: paddle/fluid/framework/details/multi_devices_graph_builder.cc [hunk 2]
```diff
@@ -170,6 +170,7 @@ size_t MultiDevSSAGraphBuilder::GetAppropriateDeviceID(
     const std::vector<std::string> &var_names) const {
   int64_t numel_sum = 0;
   for (auto var_name : var_names) {
+    if (all_vars_.find(var_name) == all_vars_.end()) continue;
```
Comment (Yancey0623) [inline paddle/fluid/framework/details/multi_devices_graph_builder.cc:173]: `continue` will cause the following `ENFORCE_NOT_NULL` doesn't work, are we sure do that?

Comment (panyx0718) [reply paddle/fluid/framework/details/multi_devices_graph_builder.cc:173]: only var with var_desc are filled into the all_vars_ now. So the enforce_not_null is not necessary now.

**Pull request merged**